rasa/nlu/data_augmentation/augmenter.py

```diff
@@ -0,0 +1,623 @@
+        classification_report=classification_report,
+        training_intents=nlu_training_data.intents,
+        pooled_intents=pooled_intents,
+        output_directory=output_directory,
+    )
+
+    # Plot data augmentation summary
+    _plot_summary_report(
+        intent_summary=intent_summary, output_directory=output_directory,
+    )
+
+
+def run_data_augmentation_random_sampling(
+    nlu_training_data: TrainingData,
+    paraphrase_pool: Dict[Text, List],
+    training_data_pool: Dict[Text, List],
+    pooled_intents: Set[Text],
+    output_directory: Text,
+    config: Text,
+    nlu_evaluation_data: TrainingData,
+    classification_report: Dict[Text, Dict[Text, float]],
+    random_seed: int,
+) -> None:
+    """Runs the NLU train/test cycle with data augmentation (random sampling) and generates the reports and plots summarising the impact of data augmentation on model performance.
+
+    Args:
+        nlu_training_data: NLU training data (without data augmentation).
+        paraphrase_pool: Available paraphrases for data augmentation.
+        training_data_pool: Training examples grouped by intent.
+        pooled_intents: The intents that have been selected for data augmentation.
+        output_directory: Directory to store the output files in.
+        config: NLU model config.
+        nlu_evaluation_data: NLU evaluation data.
+        classification_report: Classification report of the model run *without* data augmentation.
+        random_seed: Random seed for sampling the paraphrases.
+    """
+    # Build augmentation pools based on random sampling
+    random_expansion = _build_random_augmentation_pool(
+        paraphrase_pool, random_seed=random_seed
+    )
+
+    # Build new augmented training data
+    augmented_training_data = _build_augmented_training_data(
+        nlu_training_data=nlu_training_data,
+        training_data_pool=training_data_pool,
+        data_augmentation_pool=random_expansion,
+    )
+
+    # Store augmented training data to file
+    rasa.shared.utils.io.create_directory(output_directory)
+    nlu_training_file = os.path.join(output_directory, "train_augmented_random.yml")
+    augmented_training_data.persist_nlu(filename=nlu_training_file)
+
+    # Run NLU train/test loop
+    intent_report = _train_test_nlu_model(
+        output_directory=output_directory,
+        nlu_training_file=nlu_training_file,
+        config=config,
+        nlu_evaluation_data=nlu_evaluation_data,
+    )
+
+    # Create data augmentation summary
+    (intent_summary, _) = _create_summary_report(
+        intent_report=intent_report,
+        classification_report=classification_report,
+        training_intents=nlu_training_data.intents,
+        pooled_intents=pooled_intents,
+        output_directory=output_directory,
+    )
+
+    # Plot data augmentation summary
+    _plot_summary_report(
+        intent_summary=intent_summary, output_directory=output_directory,
+    )
+
+
+def augment_nlu_data(
```

Comment:
Okay now I see one problem with this function - We are creating our own data structure for storing paraphrases, original training data etc. I think it would be a lot simpler if we created two `TrainingData` objects. One for original training data that has to be used(it already exists) and the other object composed of paraphrases as `Message` objects themselves. Currently the paraphrases are being passed as a list of strings which is harder to tests in pytests. Creating `Message` objects for them right away will not assume we are creating a new data structure which should be tested.